File: 2213-longest-substring-of-one-repeating-character/2213-longest-substring-of-one-repeating-character.java
class Solution {

    class Node {
        int maxLen;
        int prefixLen;
        int suffixLen;
        char firstChar;
        char lastChar;
        int len;

        Node(char c) {
            maxLen = 1;
            prefixLen = 1;
            suffixLen = 1;
            firstChar = c;
            lastChar = c;
            len = 1;
        }

        Node() {}
    }

    Node[] tree;
    char[] chars;

    public int[] longestRepeating(
        String s,
        String queryCharacters,
        int[] queryIndices
    ) {

        int n = s.length();
        int k = queryIndices.length;

        chars = s.toCharArray();
        tree = new Node[4 * n];

        // Build Segment Tree
        build(1, 0, n - 1);

        int[] ans = new int[k];

        for (int i = 0; i < k; i++) {

            int index = queryIndices[i];
            char ch = queryCharacters.charAt(i);

            // Update character
            chars[index] = ch;

            update(1, 0, n - 1, index, ch);

            // Root represents the complete string
            ans[i] = tree[1].maxLen;
        }

        return ans;
    }

    private void build(int node, int start, int end) {

        // Leaf node
        if (start == end) {
            tree[node] = new Node(chars[start]);
            return;
        }

        int mid = start + (end - start) / 2;

        build(2 * node, start, mid);
        build(2 * node + 1, mid + 1, end);

        tree[node] = merge(
            tree[2 * node],
            tree[2 * node + 1]
        );
    }

    private void update(
        int node,
        int start,
        int end,
        int index,
        char ch
    ) {

        // Leaf node
        if (start == end) {
            tree[node] = new Node(ch);
            return;
        }

        int mid = start + (end - start) / 2;

        if (index <= mid) {
            update(2 * node, start, mid, index, ch);
        } else {
            update(2 * node + 1, mid + 1, end, index, ch);
        }

        // Recalculate current node
        tree[node] = merge(
            tree[2 * node],
            tree[2 * node + 1]
        );
    }

    private Node merge(Node left, Node right) {

        Node res = new Node();

        res.len = left.len + right.len;

        res.firstChar = left.firstChar;
        res.lastChar = right.lastChar;

        // Initially, answer is either in left or right
        res.maxLen = Math.max(
            left.maxLen,
            right.maxLen
        );

        res.prefixLen = left.prefixLen;
        res.suffixLen = right.suffixLen;

        // Boundary characters are same
        if (left.lastChar == right.firstChar) {

            // A repeating sequence can cross the boundary
            res.maxLen = Math.max(
                res.maxLen,
                left.suffixLen + right.prefixLen
            );

            // Entire left segment is same character
            if (left.prefixLen == left.len) {
                res.prefixLen =
                    left.len + right.prefixLen;
            }

            // Entire right segment is same character
            if (right.suffixLen == right.len) {
                res.suffixLen =
                    right.len + left.suffixLen;
            }
        }

        return res;
    }
}

File: 2213-longest-substring-of-one-repeating-character/2213-longest-substring-of-one-repeating-character.py
class Solution:

    class Node:
        def __init__(self, ch=None):
            if ch is not None:
                self.maxLen = 1
                self.prefixLen = 1
                self.suffixLen = 1
                self.firstChar = ch
                self.lastChar = ch
                self.length = 1
            else:
                self.maxLen = 0
                self.prefixLen = 0
                self.suffixLen = 0
                self.firstChar = ''
                self.lastChar = ''
                self.length = 0

    def longestRepeating(self, s, queryCharacters, queryIndices):

        n = len(s)

        self.chars = list(s)
        self.tree = [None] * (4 * n)

        self.build(1, 0, n - 1)

        ans = []

        for i in range(len(queryIndices)):

            index = queryIndices[i]
            ch = queryCharacters[i]

            self.chars[index] = ch

            self.update(1, 0, n - 1, index, ch)

            ans.append(self.tree[1].maxLen)

        return ans

    def build(self, node, start, end):

        if start == end:
            self.tree[node] = self.Node(self.chars[start])
            return

        mid = (start + end) // 2

        self.build(2 * node, start, mid)
        self.build(2 * node + 1, mid + 1, end)

        self.tree[node] = self.merge(
            self.tree[2 * node],
            self.tree[2 * node + 1]
        )

    def update(self, node, start, end, index, ch):

        if start == end:
            self.tree[node] = self.Node(ch)
            return

        mid = (start + end) // 2

        if index <= mid:
            self.update(2 * node, start, mid, index, ch)
        else:
            self.update(2 * node + 1, mid + 1, end, index, ch)

        self.tree[node] = self.merge(
            self.tree[2 * node],
            self.tree[2 * node + 1]
        )

    def merge(self, left, right):

        res = self.Node()

        res.length = left.length + right.length
        res.firstChar = left.firstChar
        res.lastChar = right.lastChar

        res.maxLen = max(
            left.maxLen,
            right.maxLen
        )

        res.prefixLen = left.prefixLen
        res.suffixLen = right.suffixLen

        if left.lastChar == right.firstChar:

            # Repeating sequence crosses the boundary
            res.maxLen = max(
                res.maxLen,
                left.suffixLen + right.prefixLen
            )

            # Entire left segment is same character
            if left.prefixLen == left.length:
                res.prefixLen = (
                    left.length + right.prefixLen
                )

            # Entire right segment is same character
            if right.suffixLen == right.length:
                res.suffixLen = (
                    right.length + left.suffixLen
                )

        return res